pkg/gateway/routeutils/listener_attachment_helper.go

```
@@ -104,3 +123,35 @@ func (attachmentHelper *listenerAttachmentHelperImpl) kindCheck(listener gwv1.Li
 	}
 	return allowedRoutes.Has(route.GetRouteKind())
 }
+
+func (attachmentHelper *listenerAttachmentHelperImpl) hostnameCheck(listener gwv1.Listener, route preLoadRouteDescriptor) (bool, error) {
+	// A route can attach to listener if it does not have hostname or listener does not have hostname
+	if listener.Hostname == nil || len(route.GetHostnames()) == 0 {
+		return true, nil
+	}
+
+	// validate listener hostname, return if listener hostname is not valid
+	isListenerHostnameValid, err := IsHostNameInValidFormat(string(*listener.Hostname))
+	if err != nil {
+		attachmentHelper.logger.Error(err, "listener hostname is not valid", "listener", listener.Name, "hostname", *listener.Hostname)
+		return false, err
+	}
+	if !isListenerHostnameValid {
```

No need to check for this. We should always throw error if the host name is invalid.

---

see comment below

---

Sorry I mean `IsHostNameInValidFormat` is always returning `false, err` in case of invalid hostname. So there wont be any case where its throwing `false, nil` and it should not. So only err check is sufficient to return early, we dont need this block
`if !isListenerHostnameValid {
		return false, nil
}`